### ISSUE TITLE
changeset publish does not replace workspace:* with current version in workspace before publishing 

### ISSUE THREAD

**Issue by abdes** [atlassian/changesets#432]
## Affected Packages

@changesets/apply-release-plan

and maybe others...

## Problem

@zkochan has already raised an issue with support of workspace ranges in dependencies, and he already proposed a solution for his immediate need of replacing workspace:x.y.z with x.y.z and it's working. However the case of "workspace:*" is not taken care of yet. The documented behavior in all package managers is for it to be replaced with the current version in the workspace.

changeset publish does not do that resulting in the NPM package being published with 'workspace:*' which obviously does not work.

## Proposed solution

Not sure, but @zkochan already had a comment regarding this at: https://github.com/atlassian/changesets/pull/311/files#r407058482 and to my knowledge yarn2 properly publishes such packages when invoked with 'yarn npm publish'. Not sure about pnpm.
Maybe the solution is to let the package manager handle this?

**Workaround so far:**
I am using yarn v2, so the workaround for me is ugly but produces consistent results. 
- Just stop at changeset version
- Use yarn npm publish to publish packages
- manually create git tags
- push to git

**Comment by abdes** [comment]
Additional information regarding what Yarn V2 does during the publish:

- Yarn packs the package before submitting to the registry,
- prior to packing through the yarn-plugin-pack, a hook is invoked: `beforeWorkspacePacking`
- that hooks resolves the workspace ranges and resplaces them in the generated manifest for the packing tool

https://github.com/yarnpkg/berry/blob/b9ca9d35e5b8592c5b84653122c832dc194793f5/packages/plugin-pack/sources/index.ts#L40

A possible solution would be to let the package manager, in this case yarn do the packing and publishing with its knowledge of the workspace and dependencies. However it seems that changeset only detects pnpm or npm and is not aware of yarn when publishing:
https://github.com/atlassian/changesets/blob/a57d16355ad7d67b18b768c8f79224d80afa507c/packages/cli/src/commands/publish/npm-utils.ts#L32
```ts
async function getPublishTool(cwd: string): Promise<"npm" | "pnpm"> {
  const pm = await preferredPM(cwd);
  if (!pm) return "npm";
  return pm.name === "pnpm" ? "pnpm" : "npm";
}
```
It also seems from the rest of that file that there are some strong assumptions that the publishing tool is npm or behaves like npm, rather than just an external command. Yarn may not conform to that but it remains quite simple to just delegate that task of publishing to its command line.

**Comment by zkochan** [comment]
Can yarn recursively publish all the packages that have new versions?

pnpm has the `pnpm publish -r` command that (at the moment) works better than using changesets to publish the bumped versions.  I assume Yarn also has something like that.

**Comment by Andarist** [comment]
> Maybe the solution is to let the package manager handle this?

Yes, I believe this is the best option for moving forward.

>  However it seems that changeset only detects pnpm or npm and is not aware of yarn when publishing:

Can using Yarn 2 be easily distinguished from Yarn 1? I don't know details, but Yarn 1 is explicitly not used for publishing right now because there were some problems with it.

**Comment by abdes** [comment]
> Can yarn recursively publish all the packages that have new versions?
> 
> pnpm has the `pnpm publish -r` command that (at the moment) works better than using changesets to publish the bumped versions. I assume Yarn also has something like that.

Not to my knowledge in its current version. Yarn npm publish only acts on the current workspace. It can however just gracefully ignore republish with  --tolerate-republish. So I assume, that the intent of whoever wrote that part was to just call this on every workspace and just ignore the ones whose versions was not bumped.

Note that prior to publish, you are supposed to have used yarn version plugin or something else like changesets to bump the versions in manifests that had changesets.

@zkochan I am also playing around with pnpm + rush, and I only switched to yarn today to check the differences in yarn v2. The couple of things that really struck me in yarn v2 are:
- yarn + changesets = all you need for monorepo, and if you need more, you can just write a yarn plugin or a repo tool,
- zero-install + pnp are much cleaner, node_modules is a mess with linking for debugging.
- pnpm is cool but not enough on its own. monorepo tools are bloated with useless crap, redo what pnpm/yarn do better and integrate poorly with pnpm workspaces and yarn workspaces. I don't know if there is an alternative way to reduce that node_modules mess with pnpm and linking.

**Comment by abdes** [comment]
> > Maybe the solution is to let the package manager handle this?
> 
> Yes, I believe this is the best option for moving forward.
> 
> > However it seems that changeset only detects pnpm or npm and is not aware of yarn when publishing:
> 
> Can using Yarn 2 be easily distinguished from Yarn 1? I don't know details, but Yarn 1 is explicitly not used for publishing right now because there were some problems with it.

Well it can definitely be distinguished but not with just 'preferred-pm' module. I believe that module could use more precise heuristics to detect yarn v2 from yarn v1. For example the presence of `.yarn/releases/yarn-xxxx.js` or ultimately it's a yarn and then we do a `yarn --version` :-)

BTW, for the `workspace:` problem, ultimately we can simply request yarn v2 to pack the workspace in a tgz, and slightly adjust the npm/pnpm command line for publishing to use the packed output. That's it. The rest gets taken care of by changesets as usual.

**Comment by Andarist** [comment]
> BTW, for the workspace: problem, ultimately we can simply request yarn v2 to pack the workspace in a tgz, and slightly adjust the npm/pnpm command line for publishing to use the packed output. That's it. The rest gets taken care of by changesets as usual.

Seems like a more complex thing to do (with no benefits?). If we could just offload both packing and publishing to a single command then this would make our implementation easier, right?

**Comment by abdes** [comment]
@Andarist I'm always an advocate for separation of concerns. In this case, leaving the package manager do what it is intended to do is the right approach. If something is missing in yarn/npm/pnpm to make things work flawlessly for publishing before changesets take the control back for tagging the release, we should just push for it to happen there.

The second thing that comes to my mind as a suggestion is to evaluate the positioning of the yarn version plugin with regard to changesets. That plugin is justified from a yarn point of view as it does handle something that the PM knows better than anyone else: dependency versions in a workspace. At the same time, changesets does something else with change logs that the PM has nothing to do with. There may be some synergies here and some collaboration to happen between the PM for version management, identifying the changes in a workspace, and using that information to produce a workspace friendly changelog, handle git tagging, handle pre-releases etc...

Such a thing could also be used with PNPM as well if it publishes similar APIs and allows for plugin architecture that hooks into the PM lifecycle.

Ultimately, the end goal should be that PM + changesets + few PM plugins => monorepos that work everywhere, in dev environments, in CI, **with vscode**, in docker, offline, ... without having to hack the monorepo tools (i.e. lerna, rush, nx,...). to bypass/eject/... things they do which circumvent what the underlying PM usually does. It should be easier today that both yarn and pnpm bring builtin support for workspaces.

**Comment by abdes** [comment]
Tagging this issue from yarn berry which is of extreme interest to changesets imho :-)
https://github.com/yarnpkg/berry/issues/1510

**Comment by merceyz** [comment]
> Can using Yarn 2 be easily distinguished from Yarn 1? I don't know details, but Yarn 1 is explicitly not used for publishing right now because there were some problems with it.

@Andarist Yes, if the lockfile contains `__metadata` it's v2, otherwise it's v1

**Comment by abdes** [comment]
Any plan to fix this or coordinate with yarn berry?

**Comment by Andarist** [comment]
@abdes yes, we definitely want to fix that - sorry for not being responsive lately for this issue. Having a lot on my plate.

Would you be willing to work on fixing this? As discussed - I believe we should switch to using appropriate package managers for publishing, but we should keep using npm for Yarn 1 repositories. This would be the easiest solution for us, from what I understand. 

Q: Is everything that is needed for this to work with Yarn 2 included in their core? Or are some parts of this covered by their plugins and people have to opt-in?

**Comment by abdes** [comment]
@Andarist I don't wanna rush this tbh :-). I saw that you have started the discussion with the lead maintainer of berry and I think this is the right thing to do first. He is extremely open and supportive, so I suggest that the first step should be to agree with @arcanis on how this would work. I feel that he is willing to consider a solution that perfectly integrates changeset with berry and that's what we want :-)

**Comment by abdes** [comment]
> > Can using Yarn 2 be easily distinguished from Yarn 1? I don't know details, but Yarn 1 is explicitly not used for publishing right now because there were some problems with it.
> 
> @Andarist Yes, if the lockfile contains `__metadata` it's v2, otherwise it's v1

Just spent some time on a package manager detection library: https://www.npmjs.com/package/@lerepo/detect-package-manager,

There is a CLI package with it for a quick test if you wanna check it out in your environemnt. https://www.npmjs.com/package/@lerepo/detect-package-manager-cli

It detects npm, pnpm and yarn, and provides the version installed in the system. Note that when using berry, the version of yarn is the one applied in the project/workspace and not the global one. So you may have a yarn v1 installed globally but a yarn workspace setup with yarn v2.1.1. The detection will return 2.1.1 as a version and not the global yarn. This is already something implemented in yarn itself.

If the project is using yarn v1, then the version will be that of the globally installed yarn.

**Comment by okikio** [comment]
Is this still in development?

**Comment by Andarist** [comment]
@okikio it is - in a sense that I intend to work on Changesets. It isn't in a sense that I don't have much free time lately. This is what can happen in the under-funded OSS. If your company relies on Changesets, or on the OSS in general - they might consider sponsoring projects like this to improve the situation. It should be a form of insurance/risk mitigation policies for companies.

**Comment by okikio** [comment]
Thanks for your response, I'll keep this in mind

**Comment by brianespinosa** [comment]
@Andarist I just went to go sponsor because I thought I missed a link to do so on the landing page, but I did not see anything there. I think it would be great to set up some sort of sponsorship ability for this project that people could use. I know there are a number of options out there today for setting that up that are fairly low effort.

I did see you personally have sponsorship set up on your profile which I did, but I think it would be good to set up this actual repo for that as well. Maybe that is not possible since it is under the Atlassian org?

**Comment by Andarist** [comment]
> @Andarist I just went to go sponsor

Thank you ❤️ If you need any issue to be prioritized - let me know, I can't promise anything but I can try to look into it.

> I did see you personally have sponsorship set up on your profile which I did, but I think it would be good to set up this actual repo for that as well. Maybe that is not possible since it is under the Atlassian org?

Yeah, this is a tad complicated because of that at the moment. I don't really want to enter that territory to avoid potential complications.

**Comment by brianespinosa** [comment]
For anyone else landing here, the temporary solution to this for us is to overwrite these versions directly in between the `changeset version` and `changeset publish` steps. We do not check these changes in to version control. They are only for publishing packages correctly. We throw the changes away after `changeset publish` with `git checkout -- "*.json"`

You may need to tweak this javascript for your team's needs. We are explicitly looking inside `dependencies`, `peerDependencies` and `devDependencies` and overwriting whatever the current version is with the current workspace version. We do not check at all if the `workspace:<operator>` syntax is. This basically resolves to us doing an `exact` install of the workspace package.

```javascript
const { readFileSync, writeFileSync } = require('fs');
const { execSync } = require('child_process');
const wsOutput = execSync('yarn workspaces list --json', {
  encoding: 'utf-8',
});

// We will only try to update dependencies in these package file types
const DEP_TYPES_TO_UPDATE = [
  'dependencies',
  'devDependencies',
  'peerDependencies',
];

// Used to look up the current package.json version at a location
const getPackageFile = (location) =>
  JSON.parse(readFileSync(`${location}/package.json`));

// Object map of entries to make version replacement easier when we publish
const workspaceVersionMap = Object.fromEntries(
  wsOutput
    // Create an array from the command response
    .split('\n')
    // Filter out any empty strings (last newline from split)
    .filter((string) => Boolean(string))
    // Parse JSON strings
    .map((string) => JSON.parse(string))
    // Create [key, value] and lookup version
    .map(({ location, name }) => [
      name,
      { location, version: getPackageFile(location).version },
    ])
);

// const replaceWorkspaceSyntax = ()

// Go replace versions in each package.json file
for (const pkg in workspaceVersionMap) {
  const { location } = workspaceVersionMap[pkg];

  // Get the packagefile data that we will use to rewrite the version updates.
  // NOTE: This data will be mutated in JS before we output to overwrite the file
  const packageFile = getPackageFile(location);

  // Used to target a package.json dependency list type and if the dependency matches one of our workspace packages,
  // we update the version to match the current workspace version in our workspace version map.
  const mutuateDeps = (depType) => {
    // Get the dependency type from the file and iterate over each dependency
    for (const dep in packageFile[depType]) {
      // if a dependency matches one of the dependencies in our workspace map
      if (Object.keys(workspaceVersionMap).includes(dep)) {
        // we mutate the package file data in memory to match the current workspace version
        packageFile[depType][dep] = workspaceVersionMap[dep].version;
      }
    }
  };

  // Go mutate the allowed dependency types
  DEP_TYPES_TO_UPDATE.map((dependencyType) => mutuateDeps(dependencyType));

  // Write our updated files
  writeFileSync(
    `${location}/package.json`,
    // Convert the file data to json
    // include the 2 space indent formatting
    // and also add an empty newline at the end
    `${JSON.stringify(packageFile, null, 2)}\n`
  );
}
```

**Comment by Andarist** [comment]
The kind donation from @brianespinosa has helped me to regain interest in Changesets and I'm looking into fixing this soon.

**Comment by Andarist** [comment]
I've created a [pull request](https://github.com/atlassian/changesets/pull/674) that attempts to fix the issues mentioned here. I will setup snapshot releases for this PR soon-ish but in the meantime I would love to ask if there are any people willing to test this stuff out in their projects.

**Comment by brianespinosa** [comment]
@Andarist I am just getting back into the office from being out and trying not to look at any code. Let me kick the tires on that PR late this week to see how it works. Our use case is definitely more limited than others, but I can at least validate how it is working on our end.

**Comment by HosseinAgha** [comment]
@Andarist, Eagerly waiting for those snapshot releases.   
As `yarn add local-dependency` adds dependency using the `workspace:^` protocol __by default__ and manually finding and replacing the real package version is a pain.

**Comment by belgattitude** [comment]
@HosseinAgha till the feature lands, you can use this trick:

1. Run `yarn pack` to create an archive (workspace:* versions will be replaced)
2. untar the generated package.tgz  into a _release folder
3. set publishConfig.directory to point this folder
4. call `changeset publish`


Basically:

```json5
 "publishConfig": {
    "directory": "_release/package"
  },
  "scripts": {
    // build - >pack -> untar archive in publishConfig folder
    "build-publishable-release": "yarn build && rimraf ./_release && yarn pack && mkdir ./_release && tar zxvf ./package.tgz --directory ./_release && rm ./package.tgz",
}
```

As tested here: https://github.com/soluble-io/cache-interop/pull/135

**Comment by brianespinosa** [comment]
@Andarist I have yet to circle around to test this... but we do have some big version bumps coming in some of our packages this week. When we do that, I want to try this branch in our pipeline and see how it does and I can report back. It should be out by Friday so I'll report back, though again... I suspect due to our limited use case with version ranges, this PR should easily work for us.

**Comment by Pagebakers** [comment]
```bash
yarn changeset version
yarn workspaces foreach --from '@org/*' npm publish --access public
yarn changeset tag
```

Does the trick for me.

**Comment by eegli** [comment]
@belgattitude thank you so much! This is exactly what I needed.
In my case, I'm using [preconstruct](https://preconstruct.tools/) to build all packages simultaneously and combine it with yarn's `workspaces foreach` command to do the manual packing you suggested.

The release is fully automated with `changesets/action@v1` action.

If someone else uses preconstruct - here are my relevant release scripts. 
```json
  "scripts": {
    "build": "preconstruct build",
    "build:release": "yarn build && yarn forallpackages exec rm -rf _release && yarn forallpackages pack && yarn forallpackages exec \"mkdir _release && tar zxvf package.tgz --directory _release && rm package.tgz\"",
    "forallpackages": "yarn workspaces foreach --no-private --verbose",
    "release": "yarn build:release && changeset publish",
  },
```
And in the release action, simply call `yarn release`.

```yml
  - name: Create release PR or release packages
    uses: changesets/action@v1
    with:
      publish: yarn release
      commit: version-packages
      title: Version Packages
      createGithubReleases: true
    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```

Full example is [here](https://github.com/eegli/spotifly).

**Comment by konsalex** [comment]
@Pagebakers workaround works perfectly and is quite elegant, and less error prune of creating temp directories and changing the `package.json`.

Minor issue with our monorepo as it contained private packages, be sure to use the private flag if you have private packages:

```
yarn changeset version
yarn workspaces foreach --no-private --from '@neo4j-ndl/*' npm publish --access public
yarn changeset tag
```

**Comment by hanseltime** [comment]
As @konsalex points out, the workaround above releases all packages.  If you're on yarn v4, you can use the `--since` argument and override the `changeSetBaseRefs` property in your `.yarnrc.yml` so that anything that gets changed by
`changeset version` will be used to run the publish command in only those packages. (make sure you don't have other changes from other CI/CD).

```yaml
# ,yarnrc.yml

# We use this so that changesets can work in a monorepo with yarn
changesetBaseRefs: ["HEAD"]
```

```shell
yarn changeset version
yarn workspaces foreach --no-private --since npm publish --access public
yarn changeset tag
```
